File: Data Structures & Algorithms/coin-change/submission-1.rs

impl Solution {
    pub fn coin_change(coins: Vec<i32>, amount: i32) -> i32
    {
        use std::collections::*;
        // count[amount] := 1 + Min(i){ count[amount-coin[i]] }
        // ForAll(i){ count[coin[i]] := 1 }

        let coins: HashSet<_> = coins.into_iter()
            .filter(|&coin| coin <= amount)
            .collect();

        let n = amount as usize + 1;
        let mut counts = vec![None;n];

        for &coin in &coins {
            counts[coin as usize] = Some(1);
        }
        counts[0] = Some(0);

        for amount in 1..n {
            counts[amount] = coins.iter()
                .filter(|&coin| amount >= *coin as usize)
                .filter_map(|coin| counts[amount - *coin as usize])
                .map(|count| count + 1)
                .min();
        }
        return counts[amount as usize].unwrap_or(-1);
    }
}

File: Data Structures & Algorithms/palindromic-substrings/submission-1.cpp

class Solution {
public:
    int countSubstrings(string s)
    {
        auto const n = s.size();
        auto count = n;

        for (size_t i = 1; i < s.size(); ++i)
        {
            bool has_even = s[i-1] == s[i];
            bool has_odd = true;

            if (has_even) {
                count += 1;
            }

            size_t j = 1;
            for (; j < min(i,n-i) && (has_even || has_odd); ++j)
            {
                has_even &= s[i-j-1] == s[i+j];
                has_odd &= s[i-j] == s[i+j];

                if (has_even) {
                    count += 1;
                }
                if (has_odd) {
                    count += 1;
                }
            }
            if (j < n-i) {
                has_odd &= (s[i-j] == s[i+j]);

                if (has_odd) {
                    count += 1;
                }
            }
        }

        return count;
    }
};
